mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/resourcecluster/ResourceClusterGatewayAkkaImpl.java

```diff
@@ -18,38 +18,73 @@
 
 import akka.actor.ActorRef;
 import akka.pattern.Patterns;
+import com.spotify.futures.CompletableFutures;
 import io.mantisrx.common.Ack;
+import io.mantisrx.server.master.resourcecluster.RequestThrottledException;
 import io.mantisrx.server.master.resourcecluster.ResourceClusterGateway;
 import io.mantisrx.server.master.resourcecluster.ResourceClusterTaskExecutorMapper;
 import io.mantisrx.server.master.resourcecluster.TaskExecutorDisconnection;
 import io.mantisrx.server.master.resourcecluster.TaskExecutorHeartbeat;
 import io.mantisrx.server.master.resourcecluster.TaskExecutorRegistration;
 import io.mantisrx.server.master.resourcecluster.TaskExecutorStatusChange;
+import io.mantisrx.shaded.com.google.common.util.concurrent.RateLimiter;
 import java.time.Duration;
 import java.util.concurrent.CompletableFuture;
-import lombok.RequiredArgsConstructor;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Function;
 
-@RequiredArgsConstructor
 class ResourceClusterGatewayAkkaImpl implements ResourceClusterGateway {
     protected final ActorRef resourceClusterManagerActor;
     protected final Duration askTimeout;
     private final ResourceClusterTaskExecutorMapper mapper;
+    protected final RateLimiter rateLimiter;
+
+    ResourceClusterGatewayAkkaImpl(
+            ActorRef resourceClusterManagerActor,
+            Duration askTimeout,
+            ResourceClusterTaskExecutorMapper mapper,
+            int maxConcurrentRequestCount) {
+        this.resourceClusterManagerActor = resourceClusterManagerActor;
+        this.askTimeout = askTimeout;
+        this.mapper = mapper;
+
+        this.rateLimiter = RateLimiter.create(maxConcurrentRequestCount);
+    }
+
+    private <In, Out> Function<In, CompletableFuture<Out>> withThrottle(Function<In, CompletableFuture<Out>> func) {
```

Comment:
not a blocker, but have you thought about making this an interface all our routes could implement/use?